AJ 813 - Add python client documentation (#182)
DataBiosphere/terra-workspace-data-service: docs/WDS Python Client.md

# Workspace Data Service Python Client

## Set up

In terminal, simply run:
```
pip install wds-client
```
Or in jupyter notebook: 
```
!pip install wds-client
```

To upgrade since it will not happen automatically
```
pip install wds-client --upgrade
```

### Auth set up

You will need a user entity that has access to the WDS client you need to reach, you will have this access if you have access to the workspace where data that analysis is being run on is present. 

To authenticate in terminal, using azure cli with a user account 

```
az login --use-device-code
```

> This will provide a link and a code that one can visit in a browser to log in. Use the same account to authenticate with that you would lob into azure to view the subscription of where your Terra billing is deployed to. If you are unsure if you have an account that has that permission, autenticate in the context of a notebook running inside of Terra. 

Author: that is a great question - I checked and all of you do have direct subscription access, at least in dev. Not sure if that is by design or part of testing, but in theory it can be possible/granted by the owner of the subscription.

To authenticate inside a notebook in Terra, leveraging the managed identity the notebook is running under. 

```
!az login --identity --allow-no-subscriptions
cli_token = !az account get-access-token | jq .accessToken
azure_token = cli_token[0].replace('"', '')
```

### WDS Endpoint

WDS endpoint URL can be aquired from Leo APIs and should soon be available in a enviromental variable in the notebook VM. 

However, if for some reason it is not, or you are looking to write data into a different workspace data table that you also have access to (remember this will only work if the token you generated has access to that workspace), it can be aquired by calling the following code

> Replace env with "dev" or "prod", based on where your workspace is running. Workspace Id can be located either on the Data tab by clinking "data table status" or from the following env variable in the notebook: os.environ['WORKSPACE_ID'] 

Reviewer: `clicking` not `clinking` 🥂

    status_code = response.status_code
    
    if status_code != 200:
        return response.text
    
    print(f"Successfully retrieved details.")

    return json.loads(response.text)

# the response will return all proxy urls for the workspace
response = get_wds_url(current_workspaceId)
# specifically grab the wds one
wds_url = response[0]['proxyUrls']['wds']
print(wds_url)
```

Once you have bearer token and the WDS enpoint url, you are all set to create the WDS Api Client. Do note that token will expire and will be renewed every 60 minutes. 
To call WDS, you will always need to provide a version and workspace Id. 

```
import wds_client
api_client = wds_client.ApiClient(header_name='Authorization', header_value="Bearer " + azure_token)
api_client.configuration.host = wds_url

version = "v0.2"
# if running outside of Terra notebook VM, will need to set this to the desired workspace Id 
current_workspaceId = os.environ['WORKSPACE_ID'];
```

## WDS Available Client APIs

You can check out the swagger generated documentation for all available APIs [**here**](swagger-gen-docs/README.md). Below find a few specific examples for the most common use cases. 

## General WDS Info

> General WDS Info client allows to check the version of WDS that is running in a given workspace. It can also check the status of WDS, as well as its subcomponents. 

Reviewer 2: suggest - "General WDS Info client checks the version of WDS that is running"

```
generalInfo_instance = wds_client.GeneralWDSInformationApi(api_client)

# this returns a VersionResponse object, to get the build version: response.build.version
# you can also just simply print the whole response as an str: print(response)
response = generalInfo_instance.version_get()
print(response.build.version)

response = generalInfo_instance.status_get()
print(response.status)
```

> ### Records

Reviewer 2: Should we define "Record" vs "recordtype" vs schema? I am a little confused by the terminology. 

In WDS, a record is a row of a data table, and a recordtype is the name of the data table. Is recordtype is same as schema?

Author: I added a few clarifications in the Records and Shema sections, let me know if that helps!

# Workspace Data Service Python Client

## Set up

In terminal, simply run:
```
pip install wds-client
```
Or in jupyter notebook: 
```
!pip install wds-client
```

To upgrade since it will not happen automatically
```
pip install wds-client --upgrade
```

### Auth set up

> To authenticate your user entity to access a WDS data table, you must have permission to access the workspace where the data table is stored.

Reviewer: Do I have to have any access to the WDS table, or `Reader`, `Writer`, `Owner`, etc?

Reviewer 3: not as of this writing. AJ-666 and AJ-667 will be the first tickets where this becomes true (and has Leo-level multi-user apps as a prerequisite)

Author: after chatting with David more, took away this for now - only leaving instructions around authenticating as notebook identity

To authenticate in terminal, use azure cli with a user account:

```

> az login --use-device-code

Reviewer: Perhaps we should mention for now that `WDS` is only supported for `Azure` workspaces then?

Author: great point! added this right at the top

> To authenticate inside a notebook in Terra, leveraging the managed identity the notebook is running under the following commands need to be executed. These commands will acquire a token that will be used by the client to connect to WDS.

Reviewer 4: small nit! Can the first sentence be a little simpler or 2 sentences? I don't totally follow, I think the idea is we need to run these commands from the notebook to get the managed identity?

```

This will provide a link and a code that one can visit in a browser to log in. Use the same account to authenticate with that you would to log into Azure to view the subscription of where your Terra billing is deployed to. If you are unsure if you have an account that has that permission, authenticate in the context of a notebook running inside of Terra.

To authenticate inside a notebook in Terra, leveraging the managed identity the notebook is running under. 

```
!az login --identity --allow-no-subscriptions
cli_token = !az account get-access-token | jq .accessToken
azure_token = cli_token[0].replace('"', '')
```

### WDS Endpoint

> WDS endpoint URL can be acquired from Leo APIs and should soon be available in a environmental variable in the notebook VM. 

Reviewer: Is Leo a DSP-internal term? I might suggest just explaining what it is if these docs are somewhat external

Author: renamed and added a link to the repo that has a link to swagger - let me know if this looked better!

> WDS endpoint URL can be acquired from Leo APIs and should soon be available in a environmental variable in the notebook VM. 

Reviewer 3: nit: I believe it is "environment" variable, not "environmental" variable

> WDS endpoint URL can be acquired from Leo APIs and should soon be available in a environmental variable in the notebook VM. 

Reviewer 3: thinking out loud: I am not convinced we should/will be able to add the WDS endpoint URL as an environment variable in the notebook.

The logic to determine which WDS app is the correct one is simple in the most common cases, but also has a lot of edge cases. What if the first WDS app errored out and the user deployed a second one? And so forth.

In this doc, I might sidestep it entirely, and say something generic like "find the WDS app in the list of apps that Leo returns, then take its `proxyUrls.wds` value.`

> However, if for some reason it is not, or you are looking to write data into a different workspace data table that you also have access to (remember this will only work if the token you generated has access to that workspace), it can be acquired by calling the following code

Reviewer 3: what does "a different workspace data table that you also have access to" mean here? I don't understand this use case.

Author: after chatting more, removing this for now as it is more of an "advanced used scenario"

Replace env with "dev" or "prod", based on where your workspace is running. Workspace Id can be located either on the Data tab by clicking "Data Table Status" or from the following env variable in the notebook: os.environ['WORKSPACE_ID'] 
```

> def get_wds_url(workspaceId):

Reviewer: Just a thought, but should we make this a utility function somewhere in the `wds-client` package?

    """"Get url for wds."""
    
    uri = f"https://leonardo.dsde-{env}.broadinstitute.org/api/apps/v2/{workspaceId}?includeDeleted=false"
    
    headers = {"Authorization": "Bearer " + azure_token,
               "accept": "application/json"}
    
    response = requests.get(uri, headers=headers)
    status_code = response.status_code
    
    if status_code != 200:
        return response.text
    
    print(f"Successfully retrieved details.")

    return json.loads(response.text)

# the response will return all proxy urls for the workspace
response = get_wds_url(current_workspaceId)
# specifically grab the wds one

> wds_url = response[0]['proxyUrls']['wds']

Reviewer 3: as discussed in the previous comment, using `[0]` to grab the first app in the response is not going to be correct in all cases.

Author: let me know if this is what you were thinking?

print(wds_url)
```

Once you have bearer token and the WDS endpoint url, you are all set to create the WDS Api Client. Do note that token will expire and will be renewed every 60 minutes. 
To call WDS, you will always need to provide a version and workspace Id. 

```
import wds_client
api_client = wds_client.ApiClient(header_name='Authorization', header_value="Bearer " + azure_token)

> api_client.configuration.host = wds_url

Reviewer 3: thinking out loud: when we get to AJ-879 check here to see if a `api_client.configuration.httpUserAgent ` setting exists (or similar, I don't know the exact name)

version = "v0.2"
# if running outside of Terra notebook VM, will need to set this to the desired workspace Id 
current_workspaceId = os.environ['WORKSPACE_ID'];
```

## WDS Available Client APIs

You can check out the swagger generated documentation for all available APIs by following instruction to generate it [**here**](#Generate-Full-Client-Documentation). Below find a few specific examples for the most common use cases. 

## General WDS Info

> General WDS Info client checks the version of WDS that is running.

Reviewer 3: nit:
``` suggestion
General WDS Info client checks the version and current status of WDS.
```

```
generalInfo_instance = wds_client.GeneralWDSInformationApi(api_client)

# this returns a VersionResponse object, to get the build version: response.build.version
# you can also just simply print the whole response as an str: print(response)
response = generalInfo_instance.version_get()
print(response.build.version)

> 

Reviewer 3: nit: add code comment e.g.
```suggestion

# this returns the current status of WDS; if WDS is down it indicates where
# the underlying failure may be
```

record_request = wds_client.RecordRequest(attributes=dict_values);
# this will create a record with table name "testType" and record row name "testRecord"
# if you dont provide the primary_key, the operation will complete and the primary key column will be called "sys_name" by default
# in this context:
# "TestType" is the name of the Data Table
# "testRecord" is the value that will be populated for the primary key for the row to be added (left most column in the Data Table)
# Primary key is the name of the column that defines the primary key for the given Data Table
recordCreated = records_client.create_or_replace_record(current_workspaceId, version, 'testType', 'testRecord', record_request,  primary_key="column_key")
```

Example of get_records_as_tsv and then placing those into a pandas data frame:

```
import pandas as pd
records_client = wds_client.RecordsApi(api_client)
# query records that were just added
records = records_client.get_records_as_tsv(current_workspaceId, version, 'testType')
print(records)
testType = pd.read_csv(records, sep='\t')
testType.head()
```

Example of upload_tsv:

```

> tsv_file_name = "TestType_uploaded".tsv";

Reviewer 3: nit: think you have an extra quote in there:
```suggestion
tsv_file_name = "TestType_uploaded.tsv";
```

# this will create a tsv on the notebook VM or locally where this code is run
testType = testType.to_csv(tsv_file_name, sep="\t", index = False)

# Upload entity to workspace data table with name "testType_uploaded"
response = records_client.upload_tsv(current_workspaceId, version, "testType_uploaded", tsv_file_name)
print(response)
```

### Schema

Schema defines the column name and types inside a Data Table. The code below assumes that the workspace has a Data Table with the name "testType" that contains data in it. The code to create the record type and add data to it was covered in Records section.

Example of describe_record_type:

```
schema_instance = wds_client.SchemaApi(api_client)

> # get data specifics for a specific data set vs all

Reviewer 3: ```suggestion
# get data specifics for a specific table
```

ent_types = schema_instance.describe_record_type(current_workspaceId, version, 'testType')
print ("name:", ent_types.name ,"count:", ent_types.count)
```

Example of describe_all_record_types:

```
schema_instance = wds_client.SchemaApi(api_client)

> # get workspace data specifics, such as what records exist

Reviewer 3: ```suggestion
# get workspace data specifics, such as what tables exist

```

workspace_ent_type = schema_instance.describe_all_record_types(current_workspaceId, version)
for t in workspace_ent_type:
    print ("name:", t.name ,"count:", t.count)
```

### Instances

> Currently WDS only supports a single WDS instance by workspace, however in the future there will be support for multiple instances per workspace and these functions will be helpful on checking the state of running WDS instances. 

Reviewer 3: ```suggestion
Currently the Terra UI only supports a single WDS instance per workspace, however in the future there will be support for multiple instances per workspace and these functions will be helpful on checking the state of running WDS instances. 
```
